c2_components/src/mfx_c2_encoder_component.cpp

```diff
@@ -1589,16 +1589,33 @@ void MfxC2EncoderComponent::DoConfig(const std::vector<C2Param*> &params,
                 m_mfxVideoParamsConfig.mfx.FrameInfo.FrameRateExtD = 1000;
                 break;
             }
-            case kParamIndexBitrate:
+            case kParamIndexBitrate: {
+                // MFX_RATECONTROL_CQP parameter is valid only during initialization.
                 if (m_mfxVideoParamsConfig.mfx.RateControlMethod != MFX_RATECONTROL_CQP) {
                     uint32_t bitrate_value = static_cast<const C2BitrateTuning*>(param)->value;
                     if (m_state == State::STOPPED) {
                         m_mfxVideoParamsConfig.mfx.TargetKbps = bitrate_value / 1000; // Convert from bps to Kbps
-                    } else if (m_mfxVideoParamsConfig.mfx.RateControlMethod == MFX_RATECONTROL_VBR) {
+                    } else {
                         auto update_bitrate_value = [this, bitrate_value, queue_update, failures, param] () {
-                            MFX_DEBUG_TRACE("update_bitrate_value");
-                            MFX_DEBUG_TRACE_I32(bitrate_value);
+                            MFX_DEBUG_TRACE_FUNC;                            
+                            // MDSK strongly recommended to retrieve the actual working parameters by MFXVideoENCODE_GetVideoParam
+                            // function before making any changes to bitrate settings.
+                            mfxStatus mfx_res = m_mfxEncoder->GetVideoParam(&m_mfxVideoParamsConfig);
+                            if (MFX_ERR_NONE != mfx_res) {
+                                MFX_DEBUG_TRACE__mfxStatus(mfx_res);
+                                return;
+                            }
                             m_mfxVideoParamsConfig.mfx.TargetKbps = bitrate_value / 1000; // Convert from bps to Kbps
+                            // If application sets NalHrdConformance option in mfxExtCodingOption structure to ON, the only allowed bitrate control mode is VBR.
+                            // If OFF, all bitrate control modes are available.In CBR and AVBR modes the application can
+                            // change TargetKbps, in VBR mode the application can change TargetKbps and MaxKbps values.
+                            // Such change in bitrate will not result in generation of a new key-frame or sequence header.
+                            if ((m_encoderType == ENCODER_H264 || m_encoderType == ENCODER_H265) && 
+                                m_mfxVideoParamsConfig.mfx.RateControlMethod != MFX_RATECONTROL_VBR) {
+                                mfxExtCodingOption* codingOption = m_mfxVideoParamsConfig.AddExtBuffer<mfxExtCodingOption>();
+                                codingOption->NalHrdConformance = MFX_CODINGOPTION_OFF;
```

Comment:
NalHrdConformance is set enabled by default?

Reply:
I think default is MFX_CODINGOPTION_UNKNOWN. If I don't explicitly set OFF, MDSK will return MFX_ERR_INCOMPATIBLE_VIDEO_PARAM.

Reply:
How about VP9 encoder? Looks this change is not related to VP9. Is it workable for below cases?

android.media.cts.VideoCodecTest#testDynamicBitrateChangeVP9NdkCBR
android.media.cts.VideoCodecTest#testDynamicBitrateChangeVP9NdkVBR

Reply:
VP9NdkCBR and VP9NdkVBR can be passed, if HuC is on.

Reply:
what i am concerned is that if vp9 encoder cases need this change. Without this change, but huc is on, above vp9 cases can pass?

Reply:
I don't think so. The vp9 cases failed on 24th. It ran in the system with Huc on.
I deleted the check which is "m_mfxVideoParamsConfig.mfx.RateControlMethod == MFX_RATECONTROL_VBR", all CBR mode changes go to the else statement.
But why did fail with vp9 vbr mode, I can test again to find out tomorrow. Now my TGL is running complete cts.

Reply:
hi @TianmiChen , After testing in this morning, I found our program alway runs CBR mode(default value) no matter framework actually configure what is. The reason is that we are using **C2RateControlSetting** to register reflector,  but this param is defined by plugin not by framework. The framework configure bitrate control mode by **C2StreamBitrateModeTuning**. Fortunately the cts doesn't seem care which mode actually plugin use.
So these tests always go to the last else statement without my patch. That's why this change affected vp9's cases.

After this, I will consider to fix this configure problem.